drivers/kscan/kscan_npcx.c

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) 2022 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_kscan
+
+#include <sys/__assert.h>
+#include <drivers/gpio.h>
+#include <drivers/uart.h>
+#include <drivers/clock_control.h>
+#include <drivers/kscan.h>
+#include <kernel.h>
+#include <pm/device.h>
+#include <soc.h>
+#include <sys/atomic.h>
+
+#include "soc_miwu.h"
+
+#include <logging/log.h>
+#define LOG_LEVEL CONFIG_KSCAN_LOG_LEVEL
+LOG_MODULE_REGISTER(kscan_npcx);
+
+/* Driver config */
+struct kscan_npcx_config {
+	/* keyboard scan controller base address */
+	uintptr_t base;
+	/* clock configuration */
+	struct npcx_clk_cfg clk_cfg;
+	/* pinmux configuration */
+	const uint8_t alts_size;
+	const struct npcx_alt *alts_list;
+	/* Keyboard scan input (KSI) wake-up irq */
+	int irq;
+	/* Size of keyboard inputs-wui mapping array */
+	int wui_size;
+	/* Mapping table between keyboard inputs and wui */
+	struct npcx_wui wui_maps[];
+};
+
+struct kscan_npcx_data {
+	struct kscan_matrix_config *config;
+	/* Callback called when user press/release a key of matrix keyboard. */
+	kscan_callback_t callback;
+
+	atomic_t enable_scan;
+};
+
+#define HAL_INSTANCE(dev)                                                                          \
+	((struct kbs_reg *)((const struct kscan_npcx_config *)(dev)->config)->base)
```

Comment:
please avoid such kind of helpers, check what e.g. cleanup of serial drivers

Reply:
Done.